### PR TITLE
Refactored buffer class and added access methods

### PR DESCRIPTION
**Pull request by JohanMabille**
Partially replaces #13

### CODE REVIEW

File: include/sparrow/buffer.hpp [hunk 1]
```diff
@@ -15,53 +15,172 @@
...
-        struct buffer_data
+        // Should probably move to a MP util file
+        template <class T, bool is_const>
+        struct constify
+            : std::conditional<is_const, const T, T>
         {
-            using value_type = T;
-            using pointer = T*;
-            using size_type = std::size_t;
+        };
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:33]: Would you use it for all the implementations?

**Review by jjerphan — APPROVED** [review]
I just have a comment and a nitpick.

File: include/sparrow/buffer.hpp (hunk 1 above)
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:33]: I don't know yet for sure, thus the "probably" in the comment above.

File: test/test_buffer.cpp [hunk 2]
```diff
@@ -156,6 +197,53 @@ namespace sparrow
             b.resize(size2);
             CHECK_EQ(b.size(), size2);
             CHECK_EQ(b.data()[2], 2);
+
+            b.resize(size1);
+            CHECK_EQ(b.size(), size1);
+            CHECK_EQ(b.data()[2], 2);
+
+            const std::size_t size3 = 6u;
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply test/test_buffer.cpp:205]: I prefer to define variables the latest possible, just before they're used.

File: include/sparrow/buffer.hpp [hunk 3]
```diff
@@ -15,53 +15,172 @@
...
 
+#include "sparrow/iterator.hpp"
+
 namespace sparrow
 {
     namespace impl
     {
-        template <class T>
-        struct buffer_data
+        // Should probably move to a MP util file
```
Comment (Klaim) [inline include/sparrow/buffer.hpp:28]: Consider adding `TODO` so that we can easilly search this kidn of thing

**Review by Klaim — APPROVED** [review]
Mostly needs a bit more documentation but LGTM otherwise.

File: include/sparrow/buffer.hpp [hunk 4]
```diff
@@ -15,53 +15,172 @@
...
+        using self_type = buffer_iterator<T, is_const>;
+        using base_type = iterator_base
+        <
+            self_type,
+            impl::constify_t<T, is_const>,
+            std::contiguous_iterator_tag
+        >;
+        using pointer = typename base_type::pointer;
+
+        explicit buffer_iterator(pointer p = nullptr);
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:59]: > If `p == nullptr`, does this makes the iterator instance a sentinel?

Actually it's a remainder of Legacy iterators, which had to be default constructible. Let's remove the édefault" constructor.

**Pull request merged**